0300-Experimental-Xen-hacking-with-GRUB-parameters.patch

```diff
@@ -0,0 +1,366 @@
++          || (key[i] >= 'a' && key[i] <= 'z')
++          || (key[i] == '_') ) )
++      {
++        return 0;
++      }
++  }
++  return 1;
++}
++
++void
++grub_parse_xen_cmdline (void)
++{
++  word_list = grub_malloc (0);
++  current_word_len = PARSER_BASE_WORD_LEN;
++  current_word = grub_malloc (current_word_len);
++  param_keys = grub_malloc (0);
++  param_vals = grub_malloc (0);
++
++  grub_uint8_t parse_flags = 0;
++  char *cmdline = (char *)grub_xen_start_page_addr->cmd_line;
++
++  for (grub_size_t i = 0; i < grub_strlen (cmdline); i++)
++    {
++      current_char = cmdline[i];
++
++      if (parse_flags & PARSER_HIT_BACKSLASH)
++        {
++          parse_flags ^= PARSER_HIT_BACKSLASH;
++          if (!append_char_to_word (false))
++            {
++              goto cleanup;
++            }
++          continue;
++        }
++
++      if (current_char == '\\')
++        {
++          parse_flags ^= PARSER_HIT_BACKSLASH;
++          continue;
++        }
++
++      if (current_char == '\'')
++        {
++          if (parse_flags & PARSER_IN_DOUBLE_QUOTES)
++            {
++              if (!append_char_to_word (false))
++                {
++                  goto cleanup;
++                }
++              continue;
++            }
++
++          parse_flags ^= PARSER_IN_SINGLE_QUOTES;
++          continue;
++        }
++
++      if (current_char == '"')
++        {
++          if (parse_flags & PARSER_IN_SINGLE_QUOTES)
++            {
++              if (!append_char_to_word (false))
++                {
++                  goto cleanup;
++                }
++              continue;
++            }
++
++          parse_flags ^= PARSER_IN_DOUBLE_QUOTES;
++          continue;
++        }
++
++      if (current_char == ' ')
++        {
++          if (parse_flags & PARSER_IN_SINGLE_QUOTES
++              || parse_flags & PARSER_IN_DOUBLE_QUOTES)
++            {
++              if (!append_char_to_word (false))
++                {
++                  goto cleanup;
++                }
++              continue;
++            }
++
++          append_word_to_list ();
++          continue;
++        }
++
++      if (!append_char_to_word (false))
++        {
++          goto cleanup;
++        }
++    }
++
++  if (current_word_pos > 0)
++    {
++      append_word_to_list ();
++    }
++
++  param_keys = grub_realloc (param_keys, word_list_len * sizeof (char *));
++  param_vals = grub_realloc (param_vals, word_list_len * sizeof (char *));
++  for (grub_size_t i = 0; i < word_list_len; i++)
++    {
++      current_word = word_list[i];
++      current_word_len = grub_strlen (current_word) + 1;
++      char *current_word_eq_ptr = grub_strchr (current_word, '=');
++      if (current_word_eq_ptr)
++        {
++          grub_size_t eq_idx
++              = (grub_size_t)(current_word_eq_ptr - current_word);
++          grub_size_t pre_eq_len
++              = current_word_len - (current_word_len - eq_idx);
```

Comment:
This equals ... `eq_idx`

Reply:
Yeah, some of my calculations here were off initially. I fixed them in the latest patch version submitted upstream.

Reply:
ugh, you're right. Will need to fix that in the next version of the patch.